src/vmspawn/vmspawn-util.c

```diff
@@ -236,3 +266,75 @@ int find_qemu_binary(char **ret_qemu_binary) {
 
         return find_executable(qemu_arch_specific, ret_qemu_binary);
 }
+
+int vsock_fix_child_cid(unsigned *machine_cid, const char *machine, int *ret_child_sock) {
+        _cleanup_close_ int vfd = -EBADF;
+        int r, i;
+        static const uint8_t sip_key[HASH_KEY_SIZE] = { 0 };
+        struct siphash machine_hash_state, state;
+
+        /* uint64_t is required here for the ioctl call, but valid CIDs are only 32 bits */
+        uint64_t cid = *ASSERT_PTR(machine_cid);
+
+        assert(machine);
+        assert(ret_child_sock);
+
+        /* Fix the CID of the AF_VSOCK socket passed to qemu
+         *
+         * If the user has passed us a CID (machine_cid != VMADDR_CID_ANY), then attempt to bind to that CID
+         * and error if we cannot.
+         *
+         * Otherwise hash the machine name to get a random CID and attempt to bind to that.
+         * If it is occupied add more information into the hash and try again.
+         * If after 1000 attempts this hasn't worked fallback to truly random CIDs.
+         * If after another 1000 attempts this hasn't worked then give up and return EADDRNOTAVAIL.
+         */
+
+        /* don't open with O_CLOEXEC, this fd is passed to QEMU */
+        vfd = open("/dev/vhost-vsock", O_RDWR);
```

Comment:
cloexec

Reply:
this FD is passed over to qemu though, so surely we cannot mark it as close on exec as it needs to pass through execve?

Reply:
oh, please leae it O_CLOEXEC for the whole time, until the very last moment, when you drop O_CLOEXEC.

Reply:
Okay, thanks for explaining the fix.

Reply:
Fixed in 73418f96ff.